File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = [1] * len(nums)
        prefix_product = 1
        suffix_product = 1
        for i in range(len(nums)):
            product[i] *= prefix_product
            prefix_product *= nums[i]
            product[-1-i] *= suffix_product
            suffix_product *= nums[-1-i]
        return product

File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> product(nums.size(), 1);
        int prefixProduct = 1;
        int suffixProduct = 1;
        
        for (int i = 0; i < nums.size(); i++) {
            product[i] *= prefixProduct;
            prefixProduct *= nums[i];
            product[nums.size() - 1 - i] *= suffixProduct;
            suffixProduct *= nums[nums.size() - 1 - i];
        }
        return product;  
    }
};